Homework 4 by Sean Brouwer. Date: 4/12/19

Part 1:

In the past, perceptrons and expert systems had their time in the spotlight, each being considered to be the ultimate breakthrough in AI technology. History has shown that those technologies ultimately turned out to be busts, although very few people realized it at the time. A new technology has now taken over the field of AI, that being deep neural networks. I am of the belief that deep neural networks are here to stay until a better technology replaces them, but will not be used exclusively.

The main concern with neural networks is the need for data. Without a giant dataset to train on, neural networks will not perform as well as a feature-engineered linear model, or other such means. Although some applications have big data to support deep neural networks, many real-life applications are too complicated and too niche to suit them. For applications that do suit neural networks, they will beat equivalent systems nearly every time, which is why they will stay for appropriate applications.

Source: https://mc.ai/is-deep-learning-a-passing-fad/

Part 2:

In [ ]:
# Backpropagation Cycle:

# XOR = [ -AND ] AND [ OR ]
# h1 = -AND: wi1h1 = wi2h1 = -1, threshold = -1
# h2 = OR: wi1h2 = wi2h2 = 1, threshold = 1
# a = AND: wh1a = wh2a = 1, threshold = 2

# o1 = [i1 i2]*[wi1h1 wi2h1; wi1h2 wi2h2]*[wh1a; wh2a]
# o1 = [0 1]*[-1 -1; 1 1]*[1; 1]
# o1 = [1 1]*[1; 1]
# o1 = 2



Part 3:

In [31]:
import tensorflow as tf
from tensorflow import keras
import numpy as np
import matplotlib.pyplot as plt

fashion_mnist = keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

# Rescale pixel values to be between 0 and 1
train_images = train_images / 255.0
test_images = test_images / 255.0

# Build and compile model
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
])
model.compile(optimizer='adam', 
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])


Train the model on the training dataset

In [38]:
model.fit(train_images, train_labels, epochs=3)


Epoch 1/3
60000/60000 [==============================] - 10s 171us/sample - loss: 0.2361 - acc: 0.9116
Epoch 2/3
60000/60000 [==============================] - 13s 215us/sample - loss: 0.2269 - acc: 0.9152
Epoch 3/3
60000/60000 [==============================] - 13s 219us/sample - loss: 0.2220 - acc: 0.9166


Test how well the model performs on testing dataset

In [39]:
test_loss, test_acc = model.evaluate(test_images, test_labels)
print('Test accuracy:', test_acc)


10000/10000 [==============================] - 1s 74us/sample - loss: 0.3340 - acc: 0.8853
Test accuracy: 0.8853


The best results achieved through this model is 88.53% accuracy, done with 3 layers: Starting set, 128, and 10. The fitting was done with only 3 epochs to avoid overfitting.